README.md
rddtools
========================================================
[![License](http://img.shields.io/badge/license-GPLv3-brightgreen.svg?style=flat)](http://www.gnu.org/licenses/gpl-3.0.html)
[![CRAN Version](http://www.r-pkg.org/badges/version/rddtools)](http://cran.r-project.org/package=rddtools)
[![RStudio Cloud Downloads](http://cranlogs.r-pkg.org/badges/rddtools?color=brightgreen)](http://cran.rstudio.com/package=rddtools)
[![Travis-CI Build Status](https://travis-ci.org/bquast/rddtools.png?branch=master)](https://travis-ci.org/bquast/rddtools)
[![Coverage Status](https://coveralls.io/repos/bquast/rddtools/badge.svg)](https://coveralls.io/r/bquast/rddtools)

**rddtools** is a new R package under development, designed to offer a set of tools to run all the steps required for a Regression Discontinuity Design (RDD) Analysis, from primary data visualisation to discontinuity estimation, sensitivity and placebo testing. 


Installing **rddtools**
-----------------------

This github website hosts the source code. One of the easiest ways to install the package from github is by using the R package **devtools**:

```r
if (!require('devtools')) install.packages('devtools')
devtools::install_github( "bquast/rddtools" )
```

Note however the latest version of rddtools only works with R 3.0, and that you might need to install  [Rtools](http://stat.ethz.ch/CRAN/bin/windows/Rtools/) if on Windows. 


Documentation
-----------------------
The (preliminary) documentation is available in the help files directly, as well as in the *vignettes*. The vignettes can be accessed from R.

```r
vignette("rddtools")
```

rddtools: main features
-----------------------

+  Simple visualisation of the data using binned-plot: `plot()`

+ Bandwidth selection:
  + MSE-RDD bandwidth procedure of [Imbens and Kalyanaraman 2012]: `rdd_bw_ik()`
  + MSE global bandwidth procedure of [Ruppert et al 1995]: `rdd_bw_rsw()`
+ Estimation:
  +  RDD parametric estimation: `rdd_reg_lm()` This includes specifying the polynomial order, including covariates with various specifications as advocated in [Imbens and Lemieux 2008].
  +  RDD local non-parametric estimation: `rdd_reg_np()`. Can also include covariates, and allows different types of inference (fully non-parametric, or parametric approximation). 
  +  RDD generalised estimation: allows to use custom estimating functions to get the RDD coefficient. Could allow for example a probit RDD, or quantile regression.
+ Post-Estimation tools:
  + Various tools, to obtain predictions at given covariate values ( `rdd_pred()` ), or to convert to other classes, to lm ( **as.lm()** ), or to the package `np` ( `as.npreg()` ). 
  + Function to do inference with clustered data: `clusterInf()` either using a cluster covariance matrix ( **vcovCluster()** ) or by a degrees of freedom correction (as in [Cameron et al. 2008]).
+ Regression sensitivity analysis:
  + Plot the sensitivity of the coefficient with respect to the bandwith: `plotSensi()`
  + *Placebo plot* using different cutpoints: `plotPlacebo()`
+ Design sensitivity analysis:
  + McCrary test of manipulation of the forcing variable: wrapper `dens_test()` to the function `DCdensity()` from package `rdd`. 
  + Test of equal means of covariates: `covarTest_mean()`
  + Test of equal density of covariates: `covarTest_dens()`
+ Datasets
  + Contains the seminal dataset of [Lee 2008]: `house`
  + Contains functions to replicate the Monte-Carlo simulations of [Imbens and Kalyanaraman 2012]: `gen_mc_ik()`

References
-----------------------
  [Imbens and Kalyanaraman 2012]: http://ideas.repec.org/a/oup/restud/v79y2012i3p933-959.html "Imbens, G. & Kalyanaraman, K. (2012) Optimal Bandwidth Choice for the Regression Discontinuity Estimator, Review of Economic Studies, 79, 933-959"
  
  [Lee 2008]: http://ideas.repec.org/a/eee/econom/v142y2008i2p675-697.html "Lee, D. S. (2008) Randomized experiments from non-random selection in U.S. House elections, Journal of Econometrics, 142, 675-697"
  
  [Imbens and Lemieux 2008]: http://ideas.repec.org/a/eee/econom/v142y2008i2p615-635.html "Imbens, G. & Lemieux, T. (2008) Regression discontinuity designs: A guide to practice, Journal of Econometrics, Vol. 142(2), pages 615-635"
  
  [Cameron et al. 2008]: http://ideas.repec.org/a/tpr/restat/v90y2008i3p414-427.html "Cameron, Gelbach and Miller (2008) Bootstrap-Based Improvements for Inference with Clustered Errors, The Review of Economics and Statistics, Vol. 90(3), pages 414-427"
  
  [Ruppert et al 1995]: http://www.jstor.org/stable/2291516 "Ruppert, D., Sheather, S. J. and Wand, M. P. (1995). An effective bandwidth selector for local least squares regression. Journal of the American Statistical Association, 90, 1257–1270."


NEWS
RDDtools 0.5.0
=====================

* cleanup documentation


RDDtools 0.3.0
=====================

* development taken over by Bastiaan

* rename package to rddtools (from RDDtools)

* rename functions to lower case

* move package from subdir to repo root directory

* change S3class method to export for roxygen

* connect method functions with . in stead of white space

* classify default functions as RDDcoef.default etc.

* update DESCRIPTION with CRAN guidelines

* change .onLoad to .onAttach 

* remove old lyx vignette in several places

* move examples from README.Rmd to Rmd vignettes

* fix empty package dependency bug

RDDtools 0.22
===========
Updated on 21/5/14

* RDDdata: change arg z to covar, add new argument z for sharp, currently unused. 

* dens_test: work now on RDDreg, return object htest

* Multiple changes in help files

* Correct import, suggests, calls to :::


RDDtools 0.21
===========
Updated on 25/7/13

* Add new function RDDpred

* Add new model.matrix.RDDdata, preparing all output, now used by all RDDreg_np, RDDreg_lm, RDDgenre...

* Add method vcov.RDDreg, as.lm.RDDreg

* Add enw function vcovCluster2, complement doc, add M Arai, 

* Add data STAR_MHE

* Many small fixes

RDDtools 0.2
===========
Updated on 16/7/13

* Add new option to have separate or same covariates

* Add as.nprg, to convert to a np regression from package np

* Add RDDcoef, working on multiple models (lm, np, npreg). 

* Many fixes...

RDDtools 0.1
===========
Initial commit on 29/04/2013

* Initial commit, containing RDDdata, RDDreg_lm, RDDreg_np, plotSensi, plotPlacebo, etc...


R/bw_ik.R
#' Imbens-Kalyanaraman Optimal Bandwidth Calculation
#' 
#' Imbens-Kalyanaraman optimal bandwidth
#' for local linear regression in Regression discontinuity designs.
#' 
#' @param rdd_object of class rdd_data created by \code{\link{rdd_data}}
#' @param kernel The type of kernel used: either \code{triangular} or \code{uniform}. 
#' @return The optimal bandwidth
#' @references Imbens, Guido and Karthik Kalyanaraman. (2012) 'Optimal Bandwidth Choice for the regression discontinuity estimator,' 
#' Review of Economic Studies (2012) 79, 933-959
#' @seealso \code{\link{rdd_bw_rsw}} Global bandwidth selector of Ruppert, Sheather and Wand (1995)
#' @author Matthieu Stigler <\email{Matthieu.Stigler@@gmail.com}>
#' @export
#' @examples
#' data(house)
#' rd<- rdd_data(x=house$x, y=house$y, cutpoint=0)
#' rdd_bw_ik(rd)


rdd_bw_ik <- function(rdd_object, kernel = c("Triangular", "Uniform", "Normal")) {
    
    kernel <- match.arg(kernel)
    checkIsRDD(rdd_object)
    cutpoint <- getCutpoint(rdd_object)
    
    res <- rdd_bw_ik_low(X = rdd_object$x, Y = rdd_object$y, threshold = cutpoint, verbose = FALSE, type = "RES", returnBig = FALSE, 
        kernel = kernel)
    return(res)
    
}

ik_bias <- function(rdd_object, kernel = c("Triangular", "Uniform", "Normal"), bw) {
    
    kernel <- match.arg(kernel)
    checkIsRDD(rdd_object)
    cutpoint <- getCutpoint(rdd_object)
    
    resB <- rdd_bw_ik_low(X = rdd_object$x, Y = rdd_object$y, threshold = cutpoint, verbose = FALSE, type = "RES", returnBig = TRUE, 
        kernel = kernel)
    
    ## compute C1: see ik equ 5, and Fan Jijbels (1996, 3.23) is done in R with locpol, computeMu(i=2, equivKernel(TrianK, nu=0,
    ## deg=1, lower=0, upper=Inf), lower=0, upper=Inf)
    C1 <- switch(kernel, Triangular = -0.1, Uniform = -0.1666667, Normal = -0.7519384)  ## from: 
    
    ## Compute bias as in ik equ:5, note here 1/4 is outside C1
    if (missing(bw)) 
        bw <- resB$h_opt
    res <- C1 * 1/2 * bw^2 * (resB$m2_right - resB$m2_left)
    return(res)
    
}

ik_var <- function(rdd_object, kernel = c("Triangular", "Uniform", "Normal"), bw) {
    
    kernel <- match.arg(kernel)
    checkIsRDD(rdd_object)
    cutpoint <- getCutpoint(rdd_object)
    
    resB <- rdd_bw_ik_low(X = rdd_object$x, Y = rdd_object$y, threshold = cutpoint, verbose = FALSE, type = "RES", returnBig = TRUE, 
        kernel = kernel)
    
    ## compute C2: see ik equ 5, and Fan Jijbels (1996, 3.23) is done in R with locpol, computeRK(equivKernel(TrianK, nu=0, deg=1,
    ## lower=0, upper=Inf), lower=0, upper=Inf)
    C2 <- switch(kernel, Triangular = 4.8, Uniform = 4, Normal = 1.785961)  ## from: 
    
    ## Compute var as in ik equ:5,
    if (missing(bw)) 
        bw <- resB$h_op
    elem1 <- (resB$var_inh_left + resB$var_inh_right)/resB$f_cu
    elem2 <- C2/(nrow(rdd_object) * bw)
    res <- elem1 * elem2
    res
}

ik_amse <- function(rdd_object, kernel = c("Triangular", "Uniform", "Normal"), bw) {
    
    var <- ik_var(rdd_object = rdd_object, kernel = kernel, bw = bw)
    bias <- ik_bias(rdd_object = rdd_object, kernel = kernel, bw = bw)
    res <- bias^2 + var
    res
}


rdd_bw_ik_low <- function(X, Y, threshold = 0, verbose = FALSE, type = c("RES", "RES_imp", "WP"), returnBig = FALSE, kernel = c("Triangular", 
    "Uniform", "Normal")) {
    
    type <- match.arg(type)
    kernel <- match.arg(kernel)
    
    
    N <- length(X)
    N_left <- sum(X < threshold, na.rm = TRUE)
    N_right <- sum(X >= threshold, na.rm = TRUE)
    
    
    ########## STEP 1
    
    ## Silverman bandwidth
    h1 <- 1.84 * sd(X) * N^(-1/5)
    if (verbose) 
        cat("\n-h1:", h1)
    
    ## f(cut)
    isIn_h1_left <- X >= (threshold - h1) & X < threshold
    isIn_h1_right <- X >= threshold & X <= (threshold + h1)
    
    NisIn_h1_left <- sum(isIn_h1_left, na.rm = TRUE)
    NisIn_h1_right <- sum(isIn_h1_right, na.rm = TRUE)
    if (verbose) 
        cat("\n-N left /right:", NisIn_h1_left, NisIn_h1_right)
    
    
    f_cut <- (NisIn_h1_left + NisIn_h1_right)/(2 * N * h1)
    if (verbose) 
        cat("\n-f(threshold):", f_cut)
    
    ## Variances : Equ (13)
    
    var_inh_left <- var(Y[isIn_h1_left], na.rm = TRUE)
    var_inh_right <- var(Y[isIn_h1_right], na.rm = TRUE)
    
    # problem with working pap0er: Equ 4.9 is different!
    if (type == "WP") {
        denom <- 1/(NisIn_h1_left + NisIn_h1_right)
        var_inh_global <- denom * ((NisIn_h1_left - 1) * var_inh_left + (NisIn_h1_right - 1) * var_inh_right)
    }
    
    if (verbose) {
        cat("\n-Sigma^2 left:", var_inh_left, "\n-Sigma^2 right:", var_inh_right)
    }
    ########## STEP 2
    
    
    ## Global function of order 3: Equ (14)
    reg <- lm(Y ~ I(X >= threshold) + I(X - threshold) + I((X - threshold)^2) + I((X - threshold)^3))
    m3 <- 6 * coef(reg)[5]
    if (verbose) 
        cat("\n-m3:", m3)
    
    
    ## left and right bandwidths: Equ (15)
    Ck_h2 <- 3.556702  # 7200^(1/7)
    h2_left <- Ck_h2 * (var_inh_left/(f_cut * m3^2))^(1/7) * N_left^(-1/7)
    h2_right <- Ck_h2 * (var_inh_right/(f_cut * m3^2))^(1/7) * N_right^(-1/7)
    
    if (verbose) 
        cat("\n-h2 left:", h2_left, "\n-h2 right:", h2_right)
    
    ## second derivatives right/left
    isIn_h2_left <- X >= (threshold - h2_left) & X < threshold
    isIn_h2_right <- X >= threshold & X <= (threshold + h2_right)
    
    N_h2_left <- sum(isIn_h2_left, na.rm = TRUE)
    N_h2_right <- sum(isIn_h2_right, na.rm = TRUE)
    
    reg2_left <- lm(Y ~ I(X - threshold) + I((X - threshold)^2), subset = isIn_h2_left)
    reg2_right <- lm(Y ~ I(X - threshold) + I((X - threshold)^2), subset = isIn_h2_right)
    
    m2_left <- as.numeric(2 * coef(reg2_left)[3])
    m2_right <- as.numeric(2 * coef(reg2_right)[3])
    
    if (verbose) 
        cat("\n-m2 left:", m2_left, "\n-m2 right:", m2_right)
    
    ########## STEP 3
    
    ## Regularization: Equ (16)
    if (type == "RES") {
        r_left <- (2160 * var_inh_left)/(N_h2_left * h2_left^4)
        r_right <- (2160 * var_inh_right)/(N_h2_right * h2_right^4)
    } else {
        r_left <- (2160 * var_inh_global)/(N_h2_left * h2_left^4)
        r_right <- (2160 * var_inh_global)/(N_h2_right * h2_right^4)
    }
    
    
    if (verbose) 
        cat("\n-Reg left:", r_left, "\n-Reg right:", r_right)
    
    ## Compute kernel dependent constant: (see file ~/Dropbox/HEI/rdd/Rcode/ik bandwidth/bandwidth_comput.R)
    Ck <- switch(kernel, Triangular = 3.4375, Uniform = 2.70192, Normal = 1.25864)  # is not 5.4 as in paper since our kernel is on I(|x|<1), not <1/2
    
    ## Final bandwidth: Equ (17)
    h_opt <- Ck * ((var_inh_left + var_inh_right)/(f_cut * ((m2_right - m2_left)^2 + r_left + r_right)))^(1/5) * N^(-1/5)
    names(h_opt) <- "h_opt"
    
    if (verbose) 
        cat("\n\n")
    
    ### 
    if (returnBig) {
        res <- list()
        res$h_opt <- as.numeric(h_opt)
        res$var_inh_left <- var_inh_left
        res$var_inh_right <- var_inh_right
        res$m2_right <- m2_right
        res$m2_left <- m2_left
        res$f_cut <- f_cut
        res$h2_left <- h2_left
        res$h2_right <- h2_right
    } else {
        res <- h_opt
    }
    
    return(res)
} 
